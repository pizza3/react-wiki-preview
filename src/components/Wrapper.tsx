import React, { Fragment, CSSProperties } from 'react'
import Tooltip from './Tooltip'

type Props = {
  children: string
  keyword: (string | { word: string; title: string })[]
  theme: 'light' | 'dark'
  anchorStyles: CSSProperties
  mouseEnterDelay: number
}

type States = {
  str: (string | JSX.Element)[]
}

export default class Wrapper extends React.Component<Props, States> {
  static defaultProps = {
    theme: 'light',
    anchorStyles: {},
    mouseEnterDelay: 0
  }

  state: States = {
    str: []
  }

  componentDidMount() {
    if (typeof this.props.children === 'string') {
      this.addAnchors(this.props.children)
    } else {
      console.error('The children are needed to be of type "String".')
    }
  }

  addAnchors = (textContent: string) => {
    const { theme, anchorStyles, mouseEnterDelay } = this.props
    const wordArray: string[] | [] = textContent.split(' ')
    const replacement: (string | JSX.Element)[] = wordArray.map((word, index):
      | string
      | JSX.Element => {
      const data: {} | { title: string } = this.checkKeywordExist(word)
      if ('title' in data) {
        const value = data.title
        return (
          <Fragment key={index}>
            {` `}
            <Tooltip
              theme={theme}
              value={value}
              mouseEnterDelay={mouseEnterDelay}
            >
              <a
                style={anchorStyles}
                href={`https://en.wikipedia.org/wiki/${value}`}
              >
                {word}
              </a>
            </Tooltip>
          </Fragment>
        )
      }
      return ` ${word}`
    })
    this.setState({
      str: replacement
    })
  }

  checkKeywordExist = (word: string): {} | { title: string } => {
    const { keyword } = this.props
    let objValue: {} | { title: string } = {}
    keyword.forEach((element) => {
      if (typeof element === 'object') {
        if (element.word === word) {
          objValue = element
        }
      } else if (element === word) {
        objValue = { title: word }
      }
    })
    return objValue
  }

  render() {
    return <div>{this.state.str}</div>
  }
}
