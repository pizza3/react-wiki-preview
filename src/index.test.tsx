import React from 'react'
import Wrapper from '.'
import { configure, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'

configure({
  adapter: new Adapter()
})
const text = `Horologium (Latin hōrologium, from Greek ὡρολόγιον, lit. 'an instrument
for telling the hour') is a constellation of six stars faintly visible in
the southern celestial hemisphere. It was first described by the French
astronomer Nicolas-Louis de Lacaille in 1756 and visualized by him as a
clock with a pendulum and a second hand. In 1922 the constellation was
redefined by the International Astronomical Union (IAU) as a region of the
celestial sphere containing Lacaille's stars, and has since been an IAU
designated constellation. Horologium's associated region is wholly visible
to observers south of 23°N. The constellation's brightest star—and the
only one brighter than an apparent magnitude of 4—is Alpha Horologii (at
3.85), an ageing orange giant star that has swollen to around 11 times the
diameter of the Sun. The long-period variable-brightness star, R Horologii
(4.7 to 14.3), has one of the largest variations in brightness among all
stars in the night sky visible to the unaided eye. Four star systems in
the constellation are known to have exoplanets; one—Gliese 1061—contains
an exoplanet in its habitable zone.`

const keywords = [
  { word: 'Greek', title: 'Ancient Greek' },
  'star',
  'constellation',
  'Sun',
  { word: 'celestial', title: 'Astronomical object' },
  'astronomer',
  'clock',
  'exoplanet',
  'pendulum',
  { word: 'IAU', title: 'IAU designated constellations' }
]

const textLength = text.split(' ').length

const wrapper = mount(<Wrapper keyword={keywords}>{text}</Wrapper>)

describe('Wrapper Component', () => {
  it('Creates basic snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('Checks the state after initial render', () => {
    const instance = wrapper.instance() as Wrapper
    expect(instance.state.str.length).toEqual(textLength)
  })

  it('Checks for internal methods', () => {
    const instance = wrapper.instance() as Wrapper
    const returnValue = instance.checkKeywordExist('constellation')
    expect(returnValue).toEqual({ title: 'constellation' })
    const returnValue2 = instance.checkKeywordExist('moon')
    expect(returnValue2).toEqual({})
    instance.setState({ str: [] })
    instance.addAnchors(text)
    expect(instance.state.str.length).toEqual(textLength)
  })
})
