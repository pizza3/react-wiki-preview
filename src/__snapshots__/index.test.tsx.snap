// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrapper Component Creates basic snapshot 1`] = `
<Wrapper
  anchorStyles={Object {}}
  keyword={
    Array [
      Object {
        "title": "Ancient Greek",
        "word": "Greek",
      },
      "star",
      "constellation",
      "Sun",
      Object {
        "title": "Astronomical object",
        "word": "celestial",
      },
      "astronomer",
      "clock",
      "exoplanet",
      "pendulum",
      Object {
        "title": "IAU designated constellations",
        "word": "IAU",
      },
    ]
  }
  mouseEnterDelay={0}
  theme="light"
>
  <div>
     Horologium
     (Latin
     hōrologium,
     from
     
    <Tooltip
      mouseEnterDelay={0}
      theme="light"
      value="Ancient Greek"
    >
      <Trigger
        action={
          Array [
            "hover",
          ]
        }
        afterPopupVisibleChange={[Function]}
        blurDelay={0.15}
        defaultPopupVisible={false}
        destroyPopupOnHide={false}
        focusDelay={0}
        getDocument={[Function]}
        getPopupClassNameFromAlign={[Function]}
        hideAction={Array []}
        mask={false}
        maskClosable={true}
        mouseEnterDelay={0}
        mouseLeaveDelay={0.1}
        onPopupAlign={[Function]}
        onPopupVisibleChange={[Function]}
        popup={
          <Content
            theme="light"
            value="Ancient Greek"
          />
        }
        popupAlign={
          Object {
            "offset": Array [
              0,
              10,
            ],
            "points": Array [
              "tc",
              "bc",
            ],
          }
        }
        popupClassName=""
        popupStyle={Object {}}
        prefixCls="dropdown"
        showAction={Array []}
      >
        <a
          href="https://en.wikipedia.org/wiki/Ancient Greek"
          key="trigger"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          Greek
        </a>
      </Trigger>
    </Tooltip>
     ὡρολόγιον,
     lit. 'an
     instrument
for
     telling
     the
     hour')
     is
     a
     
    <Tooltip
      mouseEnterDelay={0}
      theme="light"
      value="constellation"
    >
      <Trigger
        action={
          Array [
            "hover",
          ]
        }
        afterPopupVisibleChange={[Function]}
        blurDelay={0.15}
        defaultPopupVisible={false}
        destroyPopupOnHide={false}
        focusDelay={0}
        getDocument={[Function]}
        getPopupClassNameFromAlign={[Function]}
        hideAction={Array []}
        mask={false}
        maskClosable={true}
        mouseEnterDelay={0}
        mouseLeaveDelay={0.1}
        onPopupAlign={[Function]}
        onPopupVisibleChange={[Function]}
        popup={
          <Content
            theme="light"
            value="constellation"
          />
        }
        popupAlign={
          Object {
            "offset": Array [
              0,
              10,
            ],
            "points": Array [
              "tc",
              "bc",
            ],
          }
        }
        popupClassName=""
        popupStyle={Object {}}
        prefixCls="dropdown"
        showAction={Array []}
      >
        <a
          href="https://en.wikipedia.org/wiki/constellation"
          key="trigger"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          constellation
        </a>
      </Trigger>
    </Tooltip>
     of
     six
     stars
     faintly
     visible
     in
the
     southern
     
    <Tooltip
      mouseEnterDelay={0}
      theme="light"
      value="Astronomical object"
    >
      <Trigger
        action={
          Array [
            "hover",
          ]
        }
        afterPopupVisibleChange={[Function]}
        blurDelay={0.15}
        defaultPopupVisible={false}
        destroyPopupOnHide={false}
        focusDelay={0}
        getDocument={[Function]}
        getPopupClassNameFromAlign={[Function]}
        hideAction={Array []}
        mask={false}
        maskClosable={true}
        mouseEnterDelay={0}
        mouseLeaveDelay={0.1}
        onPopupAlign={[Function]}
        onPopupVisibleChange={[Function]}
        popup={
          <Content
            theme="light"
            value="Astronomical object"
          />
        }
        popupAlign={
          Object {
            "offset": Array [
              0,
              10,
            ],
            "points": Array [
              "tc",
              "bc",
            ],
          }
        }
        popupClassName=""
        popupStyle={Object {}}
        prefixCls="dropdown"
        showAction={Array []}
      >
        <a
          href="https://en.wikipedia.org/wiki/Astronomical object"
          key="trigger"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          celestial
        </a>
      </Trigger>
    </Tooltip>
     hemisphere.
     It
     was
     first
     described
     by
     the
     French
astronomer
     Nicolas-Louis
     de
     Lacaille
     in
     1756
     and
     visualized
     by
     him
     as
     a
clock
     with
     a
     
    <Tooltip
      mouseEnterDelay={0}
      theme="light"
      value="pendulum"
    >
      <Trigger
        action={
          Array [
            "hover",
          ]
        }
        afterPopupVisibleChange={[Function]}
        blurDelay={0.15}
        defaultPopupVisible={false}
        destroyPopupOnHide={false}
        focusDelay={0}
        getDocument={[Function]}
        getPopupClassNameFromAlign={[Function]}
        hideAction={Array []}
        mask={false}
        maskClosable={true}
        mouseEnterDelay={0}
        mouseLeaveDelay={0.1}
        onPopupAlign={[Function]}
        onPopupVisibleChange={[Function]}
        popup={
          <Content
            theme="light"
            value="pendulum"
          />
        }
        popupAlign={
          Object {
            "offset": Array [
              0,
              10,
            ],
            "points": Array [
              "tc",
              "bc",
            ],
          }
        }
        popupClassName=""
        popupStyle={Object {}}
        prefixCls="dropdown"
        showAction={Array []}
      >
        <a
          href="https://en.wikipedia.org/wiki/pendulum"
          key="trigger"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          pendulum
        </a>
      </Trigger>
    </Tooltip>
     and
     a
     second
     hand.
     In
     1922
     the
     
    <Tooltip
      mouseEnterDelay={0}
      theme="light"
      value="constellation"
    >
      <Trigger
        action={
          Array [
            "hover",
          ]
        }
        afterPopupVisibleChange={[Function]}
        blurDelay={0.15}
        defaultPopupVisible={false}
        destroyPopupOnHide={false}
        focusDelay={0}
        getDocument={[Function]}
        getPopupClassNameFromAlign={[Function]}
        hideAction={Array []}
        mask={false}
        maskClosable={true}
        mouseEnterDelay={0}
        mouseLeaveDelay={0.1}
        onPopupAlign={[Function]}
        onPopupVisibleChange={[Function]}
        popup={
          <Content
            theme="light"
            value="constellation"
          />
        }
        popupAlign={
          Object {
            "offset": Array [
              0,
              10,
            ],
            "points": Array [
              "tc",
              "bc",
            ],
          }
        }
        popupClassName=""
        popupStyle={Object {}}
        prefixCls="dropdown"
        showAction={Array []}
      >
        <a
          href="https://en.wikipedia.org/wiki/constellation"
          key="trigger"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          constellation
        </a>
      </Trigger>
    </Tooltip>
     was
redefined
     by
     the
     International
     Astronomical
     Union
     (IAU)
     as
     a
     region
     of
     the
celestial
     sphere
     containing
     Lacaille's
     stars,
     and
     has
     since
     been
     an
     IAU
designated
     constellation.
     Horologium's
     associated
     region
     is
     wholly
     visible
to
     observers
     south
     of
     23°N.
     The
     constellation's
     brightest
     star—and
     the
only
     one
     brighter
     than
     an
     apparent
     magnitude
     of
     4—is
     Alpha
     Horologii
     (at
3.85),
     an
     ageing
     orange
     giant
     
    <Tooltip
      mouseEnterDelay={0}
      theme="light"
      value="star"
    >
      <Trigger
        action={
          Array [
            "hover",
          ]
        }
        afterPopupVisibleChange={[Function]}
        blurDelay={0.15}
        defaultPopupVisible={false}
        destroyPopupOnHide={false}
        focusDelay={0}
        getDocument={[Function]}
        getPopupClassNameFromAlign={[Function]}
        hideAction={Array []}
        mask={false}
        maskClosable={true}
        mouseEnterDelay={0}
        mouseLeaveDelay={0.1}
        onPopupAlign={[Function]}
        onPopupVisibleChange={[Function]}
        popup={
          <Content
            theme="light"
            value="star"
          />
        }
        popupAlign={
          Object {
            "offset": Array [
              0,
              10,
            ],
            "points": Array [
              "tc",
              "bc",
            ],
          }
        }
        popupClassName=""
        popupStyle={Object {}}
        prefixCls="dropdown"
        showAction={Array []}
      >
        <a
          href="https://en.wikipedia.org/wiki/star"
          key="trigger"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          star
        </a>
      </Trigger>
    </Tooltip>
     that
     has
     swollen
     to
     around
     11
     times
     the
diameter
     of
     the
     Sun.
     The
     long-period
     variable-brightness
     star,
     R
     Horologii
(4.7
     to
     14.3),
     has
     one
     of
     the
     largest
     variations
     in
     brightness
     among
     all
stars
     in
     the
     night
     sky
     visible
     to
     the
     unaided
     eye.
     Four
     
    <Tooltip
      mouseEnterDelay={0}
      theme="light"
      value="star"
    >
      <Trigger
        action={
          Array [
            "hover",
          ]
        }
        afterPopupVisibleChange={[Function]}
        blurDelay={0.15}
        defaultPopupVisible={false}
        destroyPopupOnHide={false}
        focusDelay={0}
        getDocument={[Function]}
        getPopupClassNameFromAlign={[Function]}
        hideAction={Array []}
        mask={false}
        maskClosable={true}
        mouseEnterDelay={0}
        mouseLeaveDelay={0.1}
        onPopupAlign={[Function]}
        onPopupVisibleChange={[Function]}
        popup={
          <Content
            theme="light"
            value="star"
          />
        }
        popupAlign={
          Object {
            "offset": Array [
              0,
              10,
            ],
            "points": Array [
              "tc",
              "bc",
            ],
          }
        }
        popupClassName=""
        popupStyle={Object {}}
        prefixCls="dropdown"
        showAction={Array []}
      >
        <a
          href="https://en.wikipedia.org/wiki/star"
          key="trigger"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          star
        </a>
      </Trigger>
    </Tooltip>
     systems
     in
the
     
    <Tooltip
      mouseEnterDelay={0}
      theme="light"
      value="constellation"
    >
      <Trigger
        action={
          Array [
            "hover",
          ]
        }
        afterPopupVisibleChange={[Function]}
        blurDelay={0.15}
        defaultPopupVisible={false}
        destroyPopupOnHide={false}
        focusDelay={0}
        getDocument={[Function]}
        getPopupClassNameFromAlign={[Function]}
        hideAction={Array []}
        mask={false}
        maskClosable={true}
        mouseEnterDelay={0}
        mouseLeaveDelay={0.1}
        onPopupAlign={[Function]}
        onPopupVisibleChange={[Function]}
        popup={
          <Content
            theme="light"
            value="constellation"
          />
        }
        popupAlign={
          Object {
            "offset": Array [
              0,
              10,
            ],
            "points": Array [
              "tc",
              "bc",
            ],
          }
        }
        popupClassName=""
        popupStyle={Object {}}
        prefixCls="dropdown"
        showAction={Array []}
      >
        <a
          href="https://en.wikipedia.org/wiki/constellation"
          key="trigger"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          constellation
        </a>
      </Trigger>
    </Tooltip>
     are
     known
     to
     have
     exoplanets;
     one—Gliese
     1061—contains
an
     
    <Tooltip
      mouseEnterDelay={0}
      theme="light"
      value="exoplanet"
    >
      <Trigger
        action={
          Array [
            "hover",
          ]
        }
        afterPopupVisibleChange={[Function]}
        blurDelay={0.15}
        defaultPopupVisible={false}
        destroyPopupOnHide={false}
        focusDelay={0}
        getDocument={[Function]}
        getPopupClassNameFromAlign={[Function]}
        hideAction={Array []}
        mask={false}
        maskClosable={true}
        mouseEnterDelay={0}
        mouseLeaveDelay={0.1}
        onPopupAlign={[Function]}
        onPopupVisibleChange={[Function]}
        popup={
          <Content
            theme="light"
            value="exoplanet"
          />
        }
        popupAlign={
          Object {
            "offset": Array [
              0,
              10,
            ],
            "points": Array [
              "tc",
              "bc",
            ],
          }
        }
        popupClassName=""
        popupStyle={Object {}}
        prefixCls="dropdown"
        showAction={Array []}
      >
        <a
          href="https://en.wikipedia.org/wiki/exoplanet"
          key="trigger"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          exoplanet
        </a>
      </Trigger>
    </Tooltip>
     in
     its
     habitable
     zone.
  </div>
</Wrapper>
`;
